ESQL: partial failure in few KNN tests

### Description

The following tests [fail](https://buildkite.com/elastic/elasticsearch-pull-request/builds/76017#01977dc1-c695-4403-a3a1-d60f9ccaa80c) when [introducing assertions](https://github.com/elastic/elasticsearch/pull/129293) that no spec-based test succeeds with (unexpected) partial results:
* `knn-function.TestKnnWithNonPushableDisjunctionsOnComplexExpressions`
* `knn-function.TestKnnWithNonPushableDisjunctions`
* `knn-function.KnnSearchWithSimilarityOption`

In all cases, the partial failure reason is `failed to create query: Cannot invoke "java.lang.Integer.intValue()" because "this.dims" is null`.

#129293 will mute these tests, this issue is to investigate if these failures are expected or not.

Repro lines (specific to #129293, but the failures aren't):
* `./gradlew ":x-pack:plugin:esql:qa:server:multi-clusters:v9.1.0#newToOld" -Dtests.class="org.elasticsearch.xpack.esql.ccq.MultiClusterSpecIT" -Dtests.method="test {knn-function.TestKnnWithNonPushableDisjunctionsOnComplexExpressions}" -Dtests.seed=A5D8467778903339 -Dtests.bwc=true -Dtests.locale=kam -Dtests.timezone=America/Aruba -Druntime.java=24`
* `./gradlew ":x-pack:plugin:esql:qa:server:multi-clusters:v9.1.0#newToOld" -Dtests.class="org.elasticsearch.xpack.esql.ccq.MultiClusterSpecIT" -Dtests.method="test {knn-function.TestKnnWithNonPushableDisjunctions}" -Dtests.seed=A5D8467778903339 -Dtests.bwc=true -Dtests.locale=kam -Dtests.timezone=America/Aruba -Druntime.java=24`
* `./gradlew ":x-pack:plugin:esql:qa:server:multi-clusters:v9.1.0#newToOld" -Dtests.class="org.elasticsearch.xpack.esql.ccq.MultiClusterSpecIT" -Dtests.method="test {knn-function.KnnSearchWithSimilarityOption}" -Dtests.seed=A5D8467778903339 -Dtests.bwc=true -Dtests.locale=kam -Dtests.timezone=America/Aruba -Druntime.java=24`

One `_cluster` content example:
`java.lang.AssertionError: unexpected partial results: _clusters={details={remote_cluster={_shards={total=1, failed=1, successful=0, skipped=0}, took=34, indices=colors, failures=[{node=null, reason={reason=failed to create query: Cannot invoke "java.lang.Integer.intValue()" because "this.dims" is null, index_uuid=OBZX77CbS6KU2F5YeAJ21w, caused_by={reason=Cannot invoke "java.lang.Integer.intValue()" because "this.dims" is null, type=null_pointer_exception}, index=remote_cluster:colors, type=query_shard_exception}, index=remote_cluster:colors, shard=0}], status=partial}}}`

--- a/docs/changelog/131081.yaml
+++ b/docs/changelog/131081.yaml
@@ -1,0 +1,6 @@
+pr: 131081
+summary: Fix knn search error when dimensions are not set
+area: Vector Search
+type: bug
+issues:
+ - 129550

--- a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/40_knn_search.yml
+++ b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/40_knn_search.yml
@@ -670,3 +670,36 @@ setup:
           properties:
             embedding:
               type: dense_vector
+
+
+---
+"Searching with no data dimensions specified":
+  - requires:
+      cluster_features: "search.vectors.no_dimensions_bugfix"
+      reason: "Search with no dimensions bugfix"
+
+  - do:
+      indices.create:
+        index: empty-test
+        body:
+          mappings:
+            properties:
+              vector:
+                type: dense_vector
+                index: true
+
+  - do:
+      search:
+        index: empty-test
+        body:
+          fields: [ "name" ]
+          knn:
+            field: vector
+            query_vector: [ -0.5, 90.0, -10, 14.8, -156.0 ]
+            k: 3
+            num_candidates: 3
+            rescore_vector:
+              oversample: 1.5
+            similarity: 0.1
+
+  - match: { hits.total.value: 0 }

--- a/server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java
@@ -35,6 +35,7 @@ import org.apache.lucene.search.BooleanQuery;
 import org.apache.lucene.search.FieldExistsQuery;
 import org.apache.lucene.search.KnnByteVectorQuery;
 import org.apache.lucene.search.KnnFloatVectorQuery;
+import org.apache.lucene.search.MatchNoDocsQuery;
 import org.apache.lucene.search.PatienceKnnVectorQuery;
 import org.apache.lucene.search.Query;
 import org.apache.lucene.search.join.BitSetProducer;
@@ -2529,6 +2530,9 @@ public class DenseVectorFieldMapper extends FieldMapper {
                 throw new IllegalArgumentException(
                     "to perform knn search on field [" + name() + "], its mapping must have [index] set to [true]"
                 );
+            }
+            if (dims == null) {
+                return new MatchNoDocsQuery("No data has been indexed for field [" + name() + "]");
             }
             KnnSearchStrategy knnSearchStrategy = heuristic.getKnnSearchStrategy();
             return switch (getElementType()) {

--- a/server/src/main/java/org/elasticsearch/search/SearchFeatures.java
+++ b/server/src/main/java/org/elasticsearch/search/SearchFeatures.java
@@ -32,6 +32,7 @@ public final class SearchFeatures implements FeatureSpecification {
     public static final NodeFeature INT_SORT_FOR_INT_SHORT_BYTE_FIELDS = new NodeFeature("search.sort.int_sort_for_int_short_byte_fields");
     static final NodeFeature MULTI_MATCH_CHECKS_POSITIONS = new NodeFeature("search.multi.match.checks.positions");
     public static final NodeFeature BBQ_HNSW_DEFAULT_INDEXING = new NodeFeature("search.vectors.mappers.default_bbq_hnsw");
+    public static final NodeFeature SEARCH_WITH_NO_DIMENSIONS_BUGFIX = new NodeFeature("search.vectors.no_dimensions_bugfix");
 
     @Override
     public Set<NodeFeature> getTestFeatures() {
@@ -41,7 +42,8 @@ public final class SearchFeatures implements FeatureSpecification {
             RESCORER_MISSING_FIELD_BAD_REQUEST,
             INT_SORT_FOR_INT_SHORT_BYTE_FIELDS,
             MULTI_MATCH_CHECKS_POSITIONS,
-            BBQ_HNSW_DEFAULT_INDEXING
+            BBQ_HNSW_DEFAULT_INDEXING,
+            SEARCH_WITH_NO_DIMENSIONS_BUGFIX
         );
     }
 }

--- a/x-pack/plugin/esql/qa/testFixtures/src/main/resources/knn-function.csv-spec
+++ b/x-pack/plugin/esql/qa/testFixtures/src/main/resources/knn-function.csv-spec
@@ -29,8 +29,7 @@ chartreuse | [127.0, 255.0, 0.0]
 // end::knn-function-result[]
 ;
 
-# https://github.com/elastic/elasticsearch/issues/129550 - Add as an example to knn function documentation
-knnSearchWithSimilarityOption-Ignore
+knnSearchWithSimilarityOption
 required_capability: knn_function_v2
 
 from colors metadata _score 
@@ -208,8 +207,7 @@ green      | false
 maroon     | false
 ;
 
-# https://github.com/elastic/elasticsearch/issues/129550
-testKnnWithNonPushableDisjunctions-Ignore
+testKnnWithNonPushableDisjunctions
 required_capability: knn_function_v2
 
 from colors metadata _score 
@@ -225,8 +223,7 @@ lemon chiffon
 papaya whip
 ;
 
-# https://github.com/elastic/elasticsearch/issues/129550
-testKnnWithNonPushableDisjunctionsOnComplexExpressions-Ignore
+testKnnWithNonPushableDisjunctionsOnComplexExpressions
 required_capability: knn_function_v2
 
 from colors metadata _score 

